app/project-manager-shim/src/projectService/ensoRunner.ts

```diff
@@ -66,6 +95,339 @@ export class EnsoRunner implements Runner {
       })
     })
   }
+
+  /** Opens a project and starts its language server. */
+  async openProject(
+    projectPath: Path,
+    projectId: string,
+    extraEnv?: Array<[string, string]>,
+  ): Promise<LanguageServerSockets> {
+    // Check if the project is already running
+    const runningProject = this.runningProjects.get(projectId)
+    if (runningProject) {
+      // Return the existing language server sockets
+      return runningProject.sockets
+    }
+
+    // Generate a random root ID for this language server session
+    const rootId = crypto.randomUUID()
+
+    // Find available ports for the language server
+    const jsonPort = await this.findAvailablePort(DEFAULT_JSONRPC_PORT)
```

Comment:
What if `runningProject` changes while we're searching for available port, or stuck at any other `await` in this function?